### ISSUE TITLE
Multiple arguments in supense mode leads to infinite rendering

### ISSUE THREAD

**Issue by AngusFu** [vercel/swr#200]
## Way to reproduce

https://codesandbox.io/s/serene-kalam-nzkft

## Code

```js
import ReactDOM from "react-dom";
import React, { Suspense, useMemo } from "react";
import useSWR from "swr";

async function fetcher() {
  return { name: "useSWR" };
}
function Profile({ id }) {
  const state = useMemo(() => ({ id }), [id]);
  const { data } = useSWR(["/api/user", state, 3], fetcher, { suspense: true });
  return <div>hello, {data.name}</div>;
}

function App() {
  return (
    <Suspense fallback={<div>loading...</div>}>
      <Profile id={1} />
    </Suspense>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

**Comment by sergiodxa** [comment]
I had the same issue, but I thought I was doing something wrong with `useMemo`.

**Comment by AngusFu** [comment]
I think this is a react issue. Each time react tries to render the suspense children, hooks like `useMemo` `useEffect` are totally new.

It's not strange, since "Suspense for Data Fetching" is marked as an "Experimental"  feature, and is classified as a "CONCURRENT MODE (EXPERIMENTAL)" feature. 

Maybe it's to early for us to try "Suspense for Data Fetching".

![image](https://user-images.githubusercontent.com/8180186/70982485-04078e80-20f2-11ea-858d-671a06280573.png)

For now,  using `JSON.stringify` might  works  😂

```js
const _useSWR = require('swr');

function useSWR(key, fn, config) {
  let newKey = key;
  let fetcher = fn;

  // fix https://github.com/zeit/swr/issues/200
  if (Array.isArray(key)) {
    newKey = [key[0], JSON.stringify(key.slice(1))];
    fetcher = (key, args) => fn(key, ...JSON.parse(args));
  }

  return _useSWR(newKey, fetcher, config);
}
```

**Comment by quietshu** [comment]
Yeah, we've updated the *Multiple Arguments* section in the README. We don't recommend to use `useMemo` to memorize query arguments now. You should really do this

```js
const { data } = useSWR(["/api/user", id, 3], fetcher, { suspense: true });
```

instead of stringify it. 😆

> Each time react tries to render the suspense children, hooks like useMemo useEffect are totally new.

Although this is indeed the cause, it might not be a "bug" because in React's docs https://reactjs.org/docs/hooks-reference.html#usememo:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee

There's no guarantee that it will always be memorized.

**Comment by nulladdict** [comment]
Isn't this the expected behaviour?
The process is something like this:
1. Components starts rendering, fiber (F1) gets created
2. Value (V1) gets memoized into associated fiber (F1) 
3. swr gets called and suspends
4. React discards fiber F1 with it's state and subsequently V1 gets lost
5. Promise gets resolved
6. Components starts rendering again, and **new** fiber (F2) gets created
7. Value (V2) gets memoized into new fiber (F2) 
8. swr gets called, sees that V1 !== V2 and suspends

I believe this is not a bug, but a desired behaviour from react, since saving and restoring state associated with a component that threw is obviously unsafe

**Comment by nandorojo** [comment]
@shuding First off, this library is amazing. I understand that your advice is to use strings, numbers, etc. as the arguments for the `key` array. 

That said, I think there are instances where one of the items in the `key` array can't be a primitive value. For example, when making a Firestore query, my library does something like this:

```jsx
const path = 'users'
const where = ['name', '==', props.name]

useSWR([path, where], firestoreHandler)
```

This is an oversimplified example of course, but I think there will inevitably be cases like this, where you need to pass an array or dictionary as one of the items in the `key` array. 

Since `useMemo` doesn't provide semantic guarantee, as you mentioned, I'm relying on [**`useMemoOne`**](https://github.com/alexreardon/use-memo-one) instead. Does this seem like it would work as a general solution? At the very least, do you think it's safe to use if `suspense` is `false`?

Thanks so much!

**Comment by shuding** [comment]
🎉 Added support for key serialization in #1429 and released as `1.1.0-beta.0`, you don't need `useMemo` anymore:

```js
const { data } = useSWR(["/api/user", { id }, 3], fetcher);
```

**Comment by nandorojo** [comment]
🤯
